### PR TITLE
Use ROOT_LIBRARY_PATH in DynamicPath

### PR DESCRIPTION
**Pull request by ChristianTackeGSI**
ROOT's "dynamic path" has some environment variables to control it. Those environment variables have some issues:

* They are dependant on the OS (DYLD* on macOS, LD_LIBRARY_PATH on Linux, etc)
* LD_LIBRARY_PATH/etc modify the system's search path for dynamic libraries, which can result in all sorts of bad things.

We would like to have a dedicated environment variable, that is
* OS independant.
* does only affect ROOT.

Let's name it ROOT_LIBRARY_PATH (suggestion on mattermost.web.cern.ch).

It was suggested to put this into `system.rootrc` and/or `.rootrc`.

This has some issues:

* `.rootrc` is good for a per user solution. We would like to have a package level solution.
* `system.rootrc` is usually a place for the local sysadmin to modify. It could be used by a distribution package to put "defaults". But that's not really nice.
* Finally, `.rootrc` entries replace `system.rootrc` entries. So any package level configuration would be gone the moment, that the user sets `Unix.*.Root.DynamicPath`.

So getting the above mentioned environment variable to work at a package level, means to put it into TUnixSystem.cpp.

cc: @dennisklein

### CODE REVIEW

**Comment by vgvassilev** [comment]
We would also need to support this on Windows for consistency. Can you make sure that patch works on Windows too?

cc: @bellenot

**Comment by ChristianTackeGSI** [comment]
> We would also need to support this on Windows for consistency. Can you make sure that patch works on Windows too?

The current patch most likely does not work on Windows (it doesn't touch `core/winnt/*`).
I don't have access to any Windows development machines, so I can't test any possible patches. I could draft up a _suggestion_ on a patch.

Maybe we should first get to the point, what needs to be done for this to get accepted for the unix world? So that the windows implementation can take care of all that?

**Comment by bellenot** [comment]
Only the PATH variable is needed on Windows

**Comment by ChristianTackeGSI** [comment]
> Only the PATH variable is needed on Windows

Okay, that sounds to me like Windows isn't affected by this PR.

Is there anything I can do, so we get forward with this?

**Comment by vgvassilev** [comment]
@bellenot, I am confused. This PR introduces another way to express the LD_LIBRARY_PATH. IIUC, the PATH variable is the analog of LD_LIBRARY_PATH for Windows. What happens if somebody defines a variable ROOT_LIBRARY_PATH and not update the PATH? On unix it'd work just fine but fail on Windows?

**Comment by Axel-Naumann** [comment]
So I think the argument that Windows uses `%PATH` is in support of having `ROOT_LIBRARY_PATH` also for Windows. @ChristianTackeGSI would you be able to help us with that, to get this merged for 6.24?

**Comment by bellenot** [comment]
So you want to introduce `ROOT_LIBRARY_PATH` on Windows? Pointing to what? DLLs or libraries? And what for?

**Comment by Axel-Naumann** [comment]
I thought the rationale is spelled out in the original post?

>    They are dependant on the OS (DYLD* on macOS, LD_LIBRARY_PATH on Linux, etc)
>    LD_LIBRARY_PATH/etc modify the system's search path for dynamic libraries, which can result in all sorts of bad things.
>
> We would like to have a dedicated environment variable, that is
>  - OS independant.
>  -  does only affect ROOT.

Why would that exclude Windows? Either we "buy" the rationale and agree that this is a useful feature (and that would then include Windows) or we think it's not needed, but then it shouldn't be needed on any platform. What am I missing / misunderstanding?

**Comment by ChristianTackeGSI** [comment]
Thanks for revisiting this!

> So I think the argument that Windows uses `%PATH` is in support of having `ROOT_LIBRARY_PATH` also for Windows. @ChristianTackeGSI would you be able to help us with that, to get this merged for 6.24?

1. So the choosing of the name `ROOT_LIBRARY_PATH` is okayed by everyone? Just to make sure, I got it.
2. So my priority (re-)ordering is also fine?
3. I can create a patch "suggestion" for Windows. But I can't test it in any way. If that is, what you'd like me to do, I can do that, yes. Disclaimer: I can't even compile test, so it might end up not being compile-able for Windows.

**Comment by bellenot** [comment]
OK, thanks for the explanation. And don't worry, I can take care of the implementation and testing on Windows.

**Comment by bellenot** [comment]
BTW, do we have any specific test for this?

**Comment by bellenot** [comment]
So here is the diff for `TWinNTSystem.cxx`, if you want to include it in the same PR:
```
diff --git a/core/winnt/src/TWinNTSystem.cxx b/core/winnt/src/TWinNTSystem.cxx
index 71cc3be3a3..7889fc9008 100644
--- a/core/winnt/src/TWinNTSystem.cxx
+++ b/core/winnt/src/TWinNTSystem.cxx
@@ -358,24 +358,26 @@ namespace {
          dynpath = "";
       }
       if (newpath) {
-
          dynpath = newpath;
-
       } else if (dynpath == "") {
+         dynpath = gSystem->Getenv("ROOT_LIBRARY_PATH");
          TString rdynpath = gEnv ? gEnv->GetValue("Root.DynamicPath", (char*)0) : "";
          rdynpath.ReplaceAll("; ", ";");  // in case DynamicPath was extended
          if (rdynpath == "") {
             rdynpath = ".;"; rdynpath += TROOT::GetBinDir();
          }
          TString path = gSystem->Getenv("PATH");
-         if (path == "")
-            dynpath = rdynpath;
-         else {
-            dynpath = path; dynpath += ";"; dynpath += rdynpath;
+         if (!path.IsNull()) {
+            if (!dynpath.IsNull())
+               dynpath += ";";
+            dynpath += path;
+         }
+         if (!rdynpath.IsNull()) {
+            if (!dynpath.IsNull())
+               dynpath += ";";
+            dynpath += rdynpath;
          }
-
       }
-
       if (!dynpath.Contains(TROOT::GetLibDir())) {
          dynpath += ";"; dynpath += TROOT::GetLibDir();
       }
```

**Comment by ChristianTackeGSI** [comment]
Thanks to @bellenot for providing the Windows part. I have squashed it into the main commit and added a `Co-Authored-By: Bertrand Bellenot @bellenot` line (hope, that's the best way to give credit?).

> BTW, do we have any specific test for this?

If I should write tests (sounds like everything else is fine), I could need a small bit of guidance (could you point me at a test, that I could learn from / extend?).

Also rebased.

**Comment by bellenot** [comment]
@ChristianTackeGSI Thanks! And my question about test mas mostly for @Axel-Naumann

**Comment by Axel-Naumann** [comment]
It'd be good to have a test for this. This can be done through Google test in `root.git/core/base/test` or in `roottest.git/root/core`. @ChristianTackeGSI would you be able to provide us with one?

**Comment by ChristianTackeGSI** [comment]
I'll look into providing some tests. I have some idea. If I need more help, I'll speak up.

**Comment by dennisklein** [comment]
> Finally, `.rootrc` entries replace `system.rootrc` entries. So any package level configuration would be gone the moment, that the user sets `Unix.*.Root.DynamicPath`.

Just for my curiosity, if it would be possible (and if not would be desired to have):

Does the file-based rootrc-System already have the primitives (like an "include/merge other rootrc file" of sorts) to construct something like `/etc/rootrc.d/<pkgname>.rootrc` where each package could store its own rootrc file with [ap,pre]pend semantics for a search path entry like `Root.DynamicPath`?

**Comment by Axel-Naumann** [comment]
I don't think that exists. Are you suggesting we collect everything from all these files - and what is their relative priority? Or if we don't collect everything, how do we detect which "package" the current process belongs to?

Another option is for a library to modify the combined `rootrc` entries, by means of `gEnv->GetValue()`, `gEnv->SetValue()`, e.g. during static initialization.

**Comment by dennisklein** [comment]
> Are you suggesting we collect everything from all these files - and what is their relative priority?

Yes, `system.rootrc` would contain a line like `import /etc/rootrc.d/*.rootrc` or similar. Relative priority could be resolved by sorting by filename (same as shell wildcard expansion) (e.g. can be used with xxx numeric prefix and such - not perfect, but probably practical enough). Examples:
* In autofs maps one can write `+dir:/etc/auto.master.d`, or
* in apache config [`IncludeOptional conf.d/*.conf`](https://httpd.apache.org/docs/2.4/mod/core.html#include).

I mean, one could also modify the `system.rootrc` file in some postinstall/preremove pkg hooks ... but that could be tricky and require more sophisticated logic than desirable in such scripts. But then again, spack, aliBuild, and other *user-space pkg managers* install each package into their own prefix and have no or limited concepts of shared directories across packages ...

In the end, I am looking for a good way to *bake-in* some default search paths at install time and from a package that depends on root, ideally without setting up environment variables at all. But having `ROOT_LIBRARY_PATH` would already go a long way for us to be able to resolve some conflicts we have when being forced to use `LD_LIBRARY_PATH`, so I don't want to stall progress on that aspect.

**Comment by ChristianTackeGSI** [comment]
Added testsuite for the Unix part.
It works on Linux. I hope it works on other Unix-like OSes.

(rebased)

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
We really only want Google tests in root.git (I realize I did not word that clearly enough), so I will apply your test (thanks!) to our dedicated roottest.git repo. Could you split the test into a separate commit? That way I don't mess with your changes as much :-) Or if you don't mind me messing with your commits I can also split this myself - let me know.

**Comment by ChristianTackeGSI** [comment]
@Axel-Naumann Okay, split into two commits.

Hopefully that helps?

**Comment by oshadura** [comment]
@ChristianTackeGSI to actually test and avoid failures on Windows you will need to fork roottest.git (there is a bug in CI). Sorry for this!

**Comment by ChristianTackeGSI** [comment]
@oshadura can you point me (us?) to the failure?

@bellenot as you seem to be much more knowledgeable about the windows things, could you take a look?

**Comment by bellenot** [comment]
@ChristianTackeGSI As @oshadura said, the error comes from a timeout (after 4 hours) while trying to checkout [roottest](https://github.com/root-project/roottest). This is due to a bug (?) in CI/git on Windows:
```
12:24:10  * Connection #0 to host github.com left intact
12:24:10  bash: /dev/tty: No such device or address
16:23:23  Cancelling nested steps due to timeout
16:23:23  Sending interrupt signal to process
16:23:30  ^CTerminate batch job (Y/N)? 
16:23:30  ^Cscript returned exit code -1073741510
```
Simply forking [roottest](https://github.com/root-project/roottest) prevent this timeout

**Comment by ChristianTackeGSI** [comment]
Okay, I have just pressed the "fork" button on roottest.

Is it that, that you wanted me to do?

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by ChristianTackeGSI** [comment]
@bellenot and @oshadura 

Thanks for explaining this to me! That wasn't quite obvious upfront.

**Comment by bellenot** [comment]
> Okay, I have just pressed the "fork" button on roottest.
> 
> Is it that, that you wanted me to do?

Exactly

**Review by oshadura — APPROVED** [review]
Build system part looks good to me (but I will need review additional review from others involved in this PR)

**Review by pcanal — APPROVED** [review]
@ChristianTackeGSI This is indeed nice.  Could you also add to the release notes an explanation of this new feature?  Thanks.

**Comment by ChristianTackeGSI** [comment]
Should I change the target branch to 6-24-00-patches?

**Comment by ChristianTackeGSI** [comment]
@pcanal

I added a new subsection to the release notes, please review and let me know, if this is the right way. I also added my name to the contributors, hope that's okay / wanted too? (Bertrand Bellenot is already listed)

I rebased on master, but as this seems to target 6.24, I wonder, whether I should rebase on 6-24-00-patches?

**Comment by pcanal** [comment]
v6.24 is already branched and closed for new features; i.e. we are in the last few steps of the release process.

>  I also added my name to the contributors, hope that's okay / wanted too?

Of course :)

**Comment by ChristianTackeGSI** [comment]
> v6.24 is already branched and closed for new features; i.e. we are in the last few steps of the release process.

I am a bit confused. The "Milestone" for this PR [was set to "6.24/00"](https://github.com/root-project/root/pull/7031#event-4437244837)?
* If this is going into 6.24, should I rebase on that branch and target it?
* If this is going into 6.26, should I rather change the v626 release notes instead of the v624 one? (and someone should update the Milestone, I guess?)

File: README/ReleaseNotes/v624/index.md
Comment (eguiraud) [inline README/ReleaseNotes/v624/index.md:77]: just wording:

```suggestion
A new way to set ROOT's "Dynamic Path" was added: the
```

?

**Comment by pcanal** [comment]
> If this is going into 6.24, should I rebase on that branch and target it?

I had missed Axel's message.  It is indeed slated to go in 6.24.   Our usual process is to first merge in master and then backport to the v6.24 patch branch.

@ChristianTackeGSI Once you update the release note, I will merge in and then can you open a new PR against the v6.24 patch branch?

File: README/ReleaseNotes/v624/index.md
Comment (ChristianTackeGSI) [reply README/ReleaseNotes/v624/index.md]: Included. Should be resolved?

Comment (eguiraud) [reply README/ReleaseNotes/v624/index.md]: indeed, thanks

**Comment by vgvassilev** [comment]
@pcanal, just heads up before merging -- https://github.com/root-project/root/pull/7031#issuecomment-804060890

**Comment by ChristianTackeGSI** [comment]
Release notes are ready from my side.

**Comment by pcanal** [comment]
@vgvassilev thanks for the reminder.

@ChristianTackeGSI Can you remove the test commit?  Do you prefer that I add to roottest or can you add it?

**Comment by ChristianTackeGSI** [comment]
@pcanal 
> @ChristianTackeGSI Can you remove the test commit? Do you prefer that I add to roottest or can you add it?

Okay, removed the test suite commit (and rebased to master)

Could you please take care of adding it to roottest?

The test suite commit is now here, for your reference:
https://github.com/ChristianTackeGSI/root/tree/pr/root_lib_env_test
(top single commit).

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
The windows failures is unrelated (test output file deleted before being closed in an unrelated test)

**Comment by pcanal** [comment]
Thank you very much!

**Comment by ChristianTackeGSI** [comment]
Many thanks to everybody involved! :-)

**Pull request merged**
